DR-2080: Azure Snapshot Directory Storage Table Entries

snf2ye opened the pull request:
The root of this change is to implement  `tableDao.addFilesToSnapshot()` to allow us to build the virtual file directory in Azure Storage Tables as a part of the snapshot create process. 

Some notable side-effects/additional changes:
- Table Naming: We were previously assuming that all referenced tables in the directory dao were accessing a table named "dataset." We now need to be able to switch between the generic "dataset" table, and a table named specifically for a given snapshot "datarepo<SNAPSHOTID>snapshot". 
- Generic Lookup/delete/create directory methods: Many of the existing Storage Tables methods were specific to retrieving dataset information from Azure Storage Tables. I changed these from accepting "datasetId" (which is used for the partition key) to now accept "collectionId" (which could either be the dataset or snapshot id). It also now accepts the desired storage table name, since this differs between snapshots and datasets.

Things not fully tested:
* Test the caching mechanism in the addEntries method
* Setting a different StorageTableClient for snapshot vs. dataset

Future to do - places for efficiency gains:
* batchRetrieveByPath

snf2ye committed:
batch lookup by file

snf2ye committed:
Move findNewDirectoryPaths to FileMetadataUtils

snf2ye committed:
wip

snf2ye committed:
rework addEntriesToSnapshot

snf2ye committed:
dynamically set storage table name

cleanup

snf2ye committed:
update addFilesToSnapshot, add skeleton test

snf2ye committed:
rework table naming

cleanup

snf2ye committed:
full workflow works

snf2ye committed:
full working

snf2ye committed:
notes to self

snf2ye committed:
Make directory entries generic between snapshot and datasets

snf2ye committed:
fix path for snapshot lookups

snf2ye committed:
cleaning up after merge conflicts

snf2ye committed:
use same entry check for both the dataset and snapshot lookup

snf2ye committed:
test partially different path

snf2ye committed:
cleanup

remove unnecessary test

spotless

spotbugs and fix test

cleanup after merge

Fix test

snf2ye committed:
fix test for real this time

snf2ye committed:
Use string utils and test

Test prepend method

snf2ye committed:
Move everything to use StorageTableUtils for naming tables

snf2ye commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
     String temp = StringUtils.prependIfMissing(fullPath, "/");
     temp = StringUtils.removeEnd(temp, "/");
-    return ROOT_DIR_NAME + temp;
+    temp = StringUtils.prependIfMissing(temp, ROOT_DIR_NAME);
+    return temp;
```
I think this was a bug - we build the file path with the ROOT_DIR_NAME and use this endpoint during snapshot create, so it would have prepended "_dr_"twice.

pshapiro4broad replied:
It seems to me that we should be able to know in the caller which version of the API is needed. I guess there's no harm in writing it way, but it does seem imprecise to me. Unless this is a user input which hasn't been validated yet, which we don't have control over.

snf2ye replied:
That's fair. It's just used across a few lookups and by adding this, it allows us to handle both cases: The case where we've already prepended it and not.  If we've already prepended it, then it's just another sanity check that it's in the right format. I don't see the harm in this imprecision, but could be convinced otherwise!

snf2ye commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
+      }
+    }
+    return pathsToCheck;
+  }
```
Just moving this to a shared location between firestore and storage tables.

snf2ye committed:
share code

snf2ye committed:
Test utility methods

snf2ye committed:
cleanup

cleanup

cleanup

tlangs commented on src/main/java/bio/terra/service/common/azure/StorageTableUtils.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  // TODO - With DR-2127, remove this  method and add case for dataset in toTableName
+  public static String getDatasetTableName() {
```
I don't think we need these to have getter methods. Why not just make them `public static final` fields and reference directly?

snf2ye replied:
That's a great point - thank you!

tlangs commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -50,12 +55,25 @@ public static boolean tableExists(TableServiceClient tableServiceClient, String
+    if (tableHasEntries(tableServiceClient, tableName, options)) {
+      TableClient tableClient = tableServiceClient.getTableClient(tableName);
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
+      return Iterables.size(entities);
```
Unfortunately, this is a very expensive way to get the size of the entities, but its the only using the storage tables API. This is going to consume the entire contents of the returned entities just to count them. In order to reduce the network traffic, we may want to set the `select` option to only grab the row key and partition key so that we're not streaming the entire entity contents over the network just to count it.

snf2ye replied:
Good point - what I really need to know here is if there is a unique entry in the storage table given the query parameters. I just find working with the PagedIterable a bit tricky, so my new solution looks a bit gross. But, it should be more efficient.

tlangs commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -74,11 +79,13 @@ public String getPartitionKey(String prefix, String path) {
   public void createDirectoryEntry(
-      TableServiceClient tableServiceClient, FireStoreDirectoryEntry createEntry) {
+      TableServiceClient tableServiceClient,
+      String collectionId,
```
if `collectionId` is always a dataset ID or snapshot ID, why not make it a UUID here? I see that we're doing `datasetId.toString()` and `snapshotId.toString()` all over the place, so it might make sense to just take UUIDs all the way down until it actually needs to be turned into a String.

snf2ye replied:
That's a good point! I think I've changed them all - let me know if you see any I missed.

tlangs commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -225,31 +285,158 @@ private TableEntity lookupByFilePath(
+      List<String> fileIds) {
+    LRUMap<String, Boolean> pathMap = new LRUMap<>(MAX_FILTER_CLAUSES);
+    storeTopDirectory(snapshotTableServiceClient, snapshotId, datasetDirName);
+    ListUtils.partition(fileIds, MAX_FILTER_CLAUSES).stream()
+        .forEach(
```
There's no need to construct a `Stream` here. You can just do `ListUtils.partition(...).forEach(...), which will do a simple for-loop underneath.

snf2ye committed:
cleanup and add test for snapshot store top directory

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableUtils.java:
```diff
@@ -0,0 +1,48 @@
+    return FILES_TABLE_NAME;
+  }
+
+  public enum StorageTableNameSuffix {
```
Something I think we should be better about is using less redundant names for scoped definitions. In this case the enum `StorageTableNameSuffix` is inside a class named `StorageTableUtils`. So it seems like the prefix `StorageTable` is redundant and could be removed.
```suggestion
  public enum NameSuffix {
```

snf2ye replied:
Good point! It should be removed now.

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableUtils.java:
```diff
@@ -0,0 +1,48 @@
+   * @param suffix suffix for storage table Name
+   * @return A valid azure storage table name
+   */
+  public static String toTableName(String resourceId, StorageTableNameSuffix suffix) {
```
I haven't yet read how these are used, but my usual rule of thumb is that if you have a `static` method that also accepts an object, it's likely that the method could be made into a non-static method on that object. In this case I think it could be moved to `StorageTableNameSuffix`. Possibly once that's done the outer class will go away and can be replaced with the enum as a top-level definition.

snf2ye replied:
Ok, I've refactored!! Please let me know what you think!

pshapiro4broad replied:
Looks good to me!

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
   }
+
+  public List<String> findNewDirectoryPaths(
+      List<FireStoreDirectoryEntry> datasetEntries, LRUMap<String, Boolean> pathMap) {
```
I don't see a need to use the more specific type `LRUMap` here. This code doesn't need to care what type of `Map` the caller passes in.
```suggestion
      List<FireStoreDirectoryEntry> datasetEntries, Map<String, Boolean> pathMap) {
```

pshapiro4broad replied:
This is also now covered in https://broadworkbench.atlassian.net/browse/DR-2145

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
+          testPath = getDirectoryPath(testPath)) {
+
+        // check the cache
+        if (pathMap.get(testPath) == null) {
+          // not in the cache: add to checklist and a to cache
+          pathsToCheck.add(testPath);
+          pathMap.put(testPath, true);
+        }
```
I think a better way to express this is to use `computeIfAbsent()`, which has the additional feature of being thread safe.
```suggestion
        pathMap.computeIfAbsent(testPath, s -> {
          pathsToCheck.add(s);
          return true;
        });
```

snf2ye replied:
Same applies here - https://github.com/DataBiosphere/jade-data-repo/pull/1073/files#r720421651

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -77,10 +80,33 @@ public FireStoreDirectoryEntry makeDirectoryEntry(String lookupDirPath) {
+  public List<String> findNewDirectoryPaths(
+      List<FireStoreDirectoryEntry> datasetEntries, LRUMap<String, Boolean> pathMap) {
+
+    List<String> pathsToCheck = new ArrayList<>();
```
Not sure why this code uses `List` instead of `Set`, as it looks like there's no benefit in checking the same path more than once. This would only be used if the number of paths exceeds the LRU cache size, which may be unlikely, but it seems like it would better express the intent of the code.

snf2ye replied:
As I mentioned in my comment, this is not new code, it's just moved to be shared between gcp and azure. I'll create a new ticket to re-factor and add tests for this method because I want to make sure I do a non-rushed job since this will impact production code (i.e. impacts both gcp, not just azure).

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -15,6 +17,9 @@
 
   public static boolean tableHasEntries(
```
Would it make sense to move these static methods to a class that contains a `TableServiceClient`? If that object's lifetime is long enough (that is, we wouldn't end up making a new one every time a method is called), it would be a better way to organize the code.

snf2ye replied:
I've assumed that we cannot pass around the tableServiceClient due to expiration concerns. For each method in the daos, we're building the client passed on azure account info, so I don't think there is a class that has the client.

pshapiro4broad replied:
Yes, I believe it's only valid for the current call stack. But at any point in the call stack, it's valid for all calls underneath it. E.g., what if the code that created a `TableServiceClient` returned a wrapper that supported the API calls we end up making on it? The object would contain all other state we have up to that point which also provides context. I'm not suggesting you do this now, it's something to think about as a future refactor.

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/FileMetadataUtilsTest.java:
```diff
@@ -0,0 +1,37 @@
+@AutoConfigureMockMvc
+@Category(Unit.class)
+public class FileMetadataUtilsTest {
+  @Autowired FileMetadataUtils fileMetadataUtils;
```
I don't see why this needs to be autowired; does it it depend on an injectable value? Otherwise it would be quicker to instantiate a copy here directly and not use `SpringRunner`, `@AutoConfigureMockMvc` and `@SpringBootTest`.

Looking at `FileMetadataUtils` I don't see why it's a `@Component` and not a class with static methods, like other Utils classes.

snf2ye replied:
Indeed! I've changed it to be static!

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtilsTest.java:
```diff
@@ -0,0 +1,136 @@
+  private static final Logger logger = LoggerFactory.getLogger(TableServiceClientUtilsTest.class);
+  private UUID datasetId;
+  private TableServiceClient tableServiceClient;
+  private List<String> directoryEntriesToCleanup = new ArrayList<>();
```
This appears to be unused

snf2ye committed:
directly reference static variable for dataset and file table name

snf2ye committed:
keep ids in UUID form until need them in string format

snf2ye committed:
remove unneeded stream

snf2ye committed:
move toTableName to non-static method in enum

snf2ye committed:
rework naming one more time

snf2ye committed:
Make file metadata utils static

samanehsan commented on src/test/java/bio/terra/service/filedata/FileMetadataUtilsTest.java:
```diff
@@ -0,0 +1,28 @@
+public class FileMetadataUtilsTest {
+
+  @Test
+  public void makeLookupPathNoRootDir() {
```
Thanks for adding in more tests! :octocat:

samanehsan commented on src/test/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtilsTest.java:
```diff
@@ -0,0 +1,129 @@
+  private TableServiceClient tableServiceClient;
+  private String tableName;
+
+  @Autowired AzureSynapsePdao azureSynapsePdao;
+  @Autowired ConnectedOperations connectedOperations;
+  @Autowired private ConnectedTestConfiguration testConfig;
+  @Autowired DatasetService datasetService;
+  @MockBean private IamProviderInterface samService;
+  @Autowired SynapseUtils synapseUtils;
+  @Autowired AzureAuthService azureAuthService;
+  @Autowired TableDirectoryDao tableDirectoryDao;
+  @Autowired TableDao tableDao;
+  @Autowired AzureBlobStorePdao azureBlobStorePdao;
+  @Autowired FileService fileService;
+  @Autowired AzureUtils azureUtils;
```
```suggestion
  @Autowired ConnectedOperations connectedOperations;
  @Autowired private ConnectedTestConfiguration testConfig;
  @MockBean private IamProviderInterface samService;
  @Autowired AzureUtils azureUtils;
```

There are a couple of autowired components here that aren't being used

samanehsan commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDirectoryDaoConnectedTest.java:
```diff
@@ -47,10 +50,12 @@
-  private String tableName;
+  private String snapshotTableName;
 
   @Autowired AzureSynapsePdao azureSynapsePdao;
```
Ahh the multi-line suggestion isn't working for me here, but I think we only need these components:
```
  @Autowired ConnectedOperations connectedOperations;
  @Autowired private ConnectedTestConfiguration testConfig;
  @MockBean private IamProviderInterface samService;
  @Autowired TableDirectoryDao tableDirectoryDao;
  @Autowired AzureUtils azureUtils;
```

snf2ye replied:
Thanks for catching these! I'm used to things that aren't used getting greyed out in intellij, but that doesn't seem to be the case 😞

snf2ye committed:
use override for dataset and files

snf2ye committed:
remove unneeded references

snf2ye committed:
rework storage table entities to avoid computationally expensive call

samanehsan commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDirectoryDaoConnectedTest.java:
```diff
@@ -86,23 +84,44 @@ public void cleanup() throws Exception {
+    // get directories to confirm the correct ones are added
+    List<String> directories = new ArrayList();
+    directories.add("/");
+    directories.add("/_dr_");
```
The `storeTopDirectory` method will also create a directory for `dataset.getName()` if it doesn't exist so you can add that into the directories to check:

```suggestion
    directories.add(dataset.getName());
```

samanehsan replied:
Oh this only creates two directory entries so I think the first two paths are getting the same record (I updated the suggested change!)

snf2ye replied:
Oh!! Thank you for catching this! I've changed the batch retrieve to also check for unique paths - I assume we wouldn't want to return duplicates.

samanehsan replied:
No problem! Checking for unique paths is a good call!!

snf2ye committed:
fix test

snf2ye committed:
fix test

samanehsan commented on src/main/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtils.java:
```diff
@@ -50,12 +56,38 @@ public static boolean tableExists(TableServiceClient tableServiceClient, String
+    }
+    if (tableHasEntries(tableServiceClient, tableName, options)) {
+      TableClient tableClient = tableServiceClient.getTableClient(tableName);
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
```
You could also `setTop(2)` here to limit the number of entities returned by the query

snf2ye replied:
Great point! I'll make that change.

snf2ye committed:
unique batch retrieve by path

snf2ye committed:
only need to return two items here

snf2ye commented:
Latest test run: https://github.com/DataBiosphere/jade-data-repo/actions/runs/1296329527

samanehsan commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -53,139 +62,184 @@
         FireStoreDirectoryEntry.toTableEntity(partitionKey, rowKey, createEntry);
-    logger.info("Creating directory entry for {} in table {}", fullPath, TABLE_NAME);
+    logger.info("Creating directory entry for {} in table {}", fullPath, tableName);
     tableClient.createEntity(createEntryEntity);
```
Should this also be `upsertEntity`?

samanehsan commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -53,139 +62,184 @@
   }
 
+  // returns empty list if no items are found
+  public List<FireStoreDirectoryEntry> batchRetrieveById(
```
Do we still need `batchRetrieveById`? Looks like it's not being used anywhere

snf2ye replied:
Huh. Good catch - I can't remember why I thought we needed this.

samanehsan commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDaoConnectedTest.java:
```diff
@@ -0,0 +1,179 @@
+  private List<String> refIds;
+  private String loadTag;
+  private int numFilesToLoad;
+  private String targetBasePathFormat = "/test/%s/file-%s.json";
```
We may see issues with parallel test runs overwriting the directory entries since there's only one "dataset" table 😞 Could we change "test" to a uuid (or the test dataset id...although that would make it show up twice in the snapshot paths)?

snf2ye replied:
Great point!

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public String toTableName() {
+      return this.label;
```
Unnecessary `this.`. In general, we don't use `this.` in the jade codebase unless it's necessary for disambiguation, such as in a constructor or setter.
```suggestion
      return label;
```

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+      throw new InvalidParameterException(
+          "Resource Id must be provided for this storage table type.");
+    }
+    return "datarepo" + resourceId.toString().replaceAll("-", "") + this.label;
```
This is different from `IngestUtils.formatSnapshotTableName()` in that it doesn't insert a `_` between the UUID and the label. Is that difference intentional? If not, it would be better to share the code instead of duplicating it.

snf2ye replied:
Yeah, the storage table name doesn't allow underscores. And that IngestUtils method is specific to naming synapse tables (vs. Storage tables).

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+   * @return A valid azure storage table name
+   */
+  public String toTableName(UUID resourceId) {
+    if (resourceId == null) {
```
Is this only to support the call from line 57 below? Or are there other ways a `null` could arrive here? If it's only to handle line 57 below, I think it would be clearer to throw it there, instead of calling this method so it's thrown here.

Note that if this check isn't done, the code will fail anyway with a NPE on line 52, so there's only a need to explicitly check and throw if we think a user will encounter this error, and would benefit from this error message.

snf2ye replied:
Ah, that makes sense. Switching it now!

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+   */
+  public String toTableName(UUID resourceId) {
+    if (resourceId == null) {
+      throw new InvalidParameterException(
```
I'm not sure `InvalidParameterException` is the best exception to use here. That error is part of the java cryptography architecture. For a general parameter check, I think `IllegalArgumentException` would be a better choice.
```suggestion
      throw new IllegalArgumentException(
```

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+   * @return A valid azure storage table name
+   */
+  public String toTableName(UUID resourceId) {
+    if (resourceId == null) {
```
We don't use it much in our code, but `Objects` has a number of null checking APIs that I think would make null checks easier for us to write. To check for `null` and return an NPE with a custom message, you can write:
```
Objects.requireNonNull(resourceId, "Resource Id must be provided for this storage table type.");
```

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -4,19 +4,18 @@
 
-@Component
 public class FileMetadataUtils {
   public static final String ROOT_DIR_NAME = "/_dr_";
```
I think conceptually this should be `private` as it's an implementation detail and shouldn't be used directly outside of this class. If it's used in testing, you can make it `protected` and add `@VisibleForTesting`.

snf2ye replied:
protected should work

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDao.java:
```diff
@@ -1,5 +1,7 @@
 package bio.terra.service.filedata.azure.tables;
 
+import static bio.terra.service.common.azure.StorageTableName.DATASET_TABLE;
```
In general I think we should avoid using `static` imports in our code, except in test code, and there usually for easier readability for hamcrest matchers, mockito, etc.

snf2ye replied:
I thought it helped with readability even in non-test code. Is there a reason why we don't want the static imports?

pshapiro4broad commented on src/main/java/bio/terra/service/common/azure/StorageTableName.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.UUID;
+
+public enum StorageTableName {
+  SNAPSHOT_TABLE("snapshot"),
+  LOAD_HISTORY_TABLE("loadHistory"),
+  DEPENDENCIES_TABLE("dependencies"),
+  DATASET_TABLE("dataset") {
```
In the interest of not using redundant names, it seems like you can remove `_TABLE` from these names without causing confusion, as they will be used as `StorageTableName.SNAPSHOT`

snf2ye replied:
I opted to use the static imports and so we could use "SNAPSHOT_TABLE.toTableName()" throughout the code rather than "StorageTableName.SNAPSHOT.toTableName()", but it sounds like static imports are not advised? Why is this the case?

snf2ye replied:
I've reworked it to use StorageTableName.SNAPSHOT.toTableName() instead.

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtilsTest.java:
```diff
@@ -0,0 +1,115 @@
+
+    tableServiceClient.createTableIfNotExists(tableName);
+    boolean tableExistsNow = TableServiceClientUtils.tableExists(tableServiceClient, tableName);
+    assertThat("table should exist", tableExistsNow, equalTo(true));
```
If you're testing for truth, you can use `assertThat(String, boolean)` and skip the matcher.
```suggestion
    assertThat("table should exist", tableExistsNow);
```

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableServiceClientUtilsTest.java:
```diff
@@ -0,0 +1,115 @@
+    assertThat("table should exist", tableExistsNow, equalTo(true));
+
+    boolean tableNoEntries = TableServiceClientUtils.tableHasEntries(tableServiceClient, tableName);
+    assertThat("table should have no entries", tableNoEntries, equalTo(false));
```
You can use `assertThat(..., !expression)` or `assertFalse()` and avoid the matcher here too. Hmm, also the state of the bool variable doesn't seem to match its value.
```suggestion
    boolean tableHasEntries = TableServiceClientUtils.tableHasEntries(tableServiceClient, tableName);
    assertThat("table should have no entries", !tableHasEntries);
```

snf2ye committed:
code review changes

snf2ye commented:
Latest run: https://github.com/DataBiosphere/jade-data-repo/actions/runs/1303835814

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDirectoryDaoTest.java:
```diff
@@ -105,9 +110,19 @@ public void testRetrieveByFileId() {
-    FireStoreDirectoryEntry response = dao.retrieveById(tableServiceClient, FILE_ID);
-    assertEquals(response, directoryEntry);
+    try (MockedStatic<TableServiceClientUtils> utils =
+        Mockito.mockStatic(TableServiceClientUtils.class)) {
```
Was there no (easy) way to do this without using `mockStatic()`? I think it's good to avoid it if possible. It looks like it could be done by mocking `TableServiceClient`, but maybe that was too difficult to do?

snf2ye replied:
I thought it was strange that I had to do this, but I haven't figured out another way to get it to work.

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDirectoryDaoConnectedTest.java:
```diff
@@ -1,27 +1,29 @@
 
+import static bio.terra.service.common.azure.StorageTableName.DATASET_TABLE;
+import static bio.terra.service.common.azure.StorageTableName.SNAPSHOT_TABLE;
+import static bio.terra.service.filedata.azure.tables.TableServiceClientUtils.tableHasEntries;
```
In this case, I think the static import makes the code harder, not easier to read.

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDirectoryDaoConnectedTest.java:
```diff
@@ -86,23 +84,53 @@ public void cleanup() throws Exception {
+    List<FireStoreDirectoryEntry> uniqueDatasetDirectoryEntries =
+        tableDirectoryDao.batchRetrieveByPath(
+            tableServiceClient, snapshotId, snapshotTableName, nonUniqueDirectories);
+    assertThat("Only 1 unique path", uniqueDatasetDirectoryEntries.size(), equalTo(1));
```
Using hamcrest `hasSize()` is a little clearer IMO
```suggestion
    assertThat("Only 1 unique path", uniqueDatasetDirectoryEntries, hasSize(1));
```

samanehsan reviewed (approved):
Lgtm! 🎉

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/tables/TableDependencyConnectedTest.java:
```diff
@@ -1,5 +1,6 @@
 package bio.terra.service.filedata.azure.tables;
 
+import static bio.terra.service.common.azure.StorageTableName.DEPENDENCIES_TABLE;
 import static org.junit.Assert.*;
```
The google java style doesn't allow wildcard imports; not sure why `spotless` allows it here.

pshapiro4broad commented on src/main/java/bio/terra/service/tabulardata/azure/AzureStorageTablePdao.java:
```diff
@@ -32,8 +33,8 @@ private static String computeInternalLoadTag(String loadTag) {
    * <p>The table name will be the result of the dataset id passed through {@link
-   * AzureStorageTablePdao#toLoadHistoryTableNameFromUUID} Entities will be partitioned on the
-   * loadTag and their row keys will be the value of {@link BulkLoadHistoryModel#getFileId()}
+   * StorageTableName#toTableName} Entities will be partitioned on the loadTag and their row keys
```
Note that since `toTableName` is overloaded it will pick the first declared name (I think). If you want to disambiguate you can add the parameter list.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/flight/delete/DeleteFileAzureDirectoryStep.java:
```diff
@@ -27,7 +29,9 @@ public StepResult doStep(FlightContext context) throws InterruptedException {
-      boolean found = tableDao.deleteDirectoryEntry(fileId, storageAuthInfo);
+      boolean found =
+          tableDao.deleteDirectoryEntry(
+              fileId, storageAuthInfo, dataset.getId(), DATASET_TABLE.toTableName());
```
A future refactor: since the last argument to `TableDao.deleteDirectoryEntry()` is always `StorageTableName.toTableName()`, you can make `deleteDirectoryEntry()` accept `StorageTableName` instead of `String` and call `toTableName()` in the method. There may be other cases like this in the code.

snf2ye replied:
Yeah, it's a little wonky right now b/c for a lot of the methods I need to be able to handle both .toTableName() and .toTableName(resourceId). So, passing around the actual table name seemed the most straight forward solution for now. In the future, when they all accept a resource id, I can move this code closer to where the table name is used.

pshapiro4broad replied:
Sounds good!

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java:
```diff
@@ -53,139 +62,166 @@
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+    return entities.stream()
+        .map(entity -> FireStoreDirectoryEntry.fromTableEntity(entity))
```
I think this is 
```suggestion
        .map(FireStoreDirectoryEntry::fromTableEntity)
```

pshapiro4broad reviewed (approved):
Looks good, I had a few more comments, but nothing to hold up the merge for.

tlangs reviewed (approved):
LGTM once all comments are resolved!

snf2ye committed:
remove references to this

snf2ye committed:
throw directly in toTableName

snf2ye committed:
Switch to Illegal Argument exception

snf2ye committed:
require not null

snf2ye committed:
remove equals true, protected root dir name

snf2ye committed:
remove equalToFalse

snf2ye committed:
remove static import

snf2ye committed:
fix imports

snf2ye committed:
fix java docs

snf2ye committed:
refactor

snf2ye committed:
rework table name

snf2ye commented:
I think I've addressed all of the comments. 
Here's the latest run: https://github.com/DataBiosphere/jade-data-repo/runs/3794886279?check_suite_focus=true

snf2ye commented:
Re-run: https://github.com/DataBiosphere/jade-data-repo/actions/runs/1305193747

snf2ye commented:
New run: https://github.com/DataBiosphere/jade-data-repo/actions/runs/1308651205
